# Data Cleaning

## REKT Database from DEFIYIELD

In [1]:
import pandas as pd             #pandas for using dataframe and reading csv file(s)
import json
import numpy as np              #numpy for vector operations and basic maths
import matplotlib.pyplot as plt #for plotting
%matplotlib inline              
import seaborn as sns           #for making plots
import math                     #for basic math operations
import warnings
# from pandas.plotting import parallel_coordinates #for multivariate plots
warnings.filterwarnings('ignore') #ignore deprecation warnings

/Users/Tegveer/miniforge3/envs/ANLY501/lib/python3.9/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.1
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


In [2]:
#importing data

open_REKTjson = open('../../data/Raw Data/Python_REKT_Database_API/REKT_Database_Python_API.json')
load_REKTjson = json.load(open_REKTjson)
REKT_df = pd.DataFrame(load_REKTjson)

REKT_df.head(5) # visualize first 5 instances of raw data

id   project_name                                        description  \
0  3058  Terra Classic  <p><strong>Quick Summary</strong></p><p>A comp...   
1  2762      Africrypt  <p><strong>Quick Summary</strong></p><p>Ameer ...   
2  2878      PlusToken  <p><strong>Quick Summary</strong></p><p>The pe...   
3  2861         Thodex  <p><strong>Quick Summary</strong></p><p>Thodex...   
4  2735     BitConnect  <p><strong>Quick Summary</strong></p><p>BitCon...   

              name_categories  token_name  \
0                  Stablecoin  LUNC, USTC   
1                        CeFi        None   
2                        CeFi        None   
3                        CeFi        None   
4  Borrowing and Lending,CeFi         BCC   

                                  proof_archive_link technical_issue  \
0  https://twitter.com/OnChainWizard/status/15241...            None   
1                                               None            None   
2                                               None            None   
3                                               None            None   
4                                               None            None   

                                       token_address  \
0  0x7e43d25EaD96B1058f671F6690ea705BA2C7e5B9, 0x...   
1                                               None   
2                                               None   
3                                               None   
4                                               None   

                                         logo_link        date  ...  \
0  safe/files/scamDatabase/logo/62b31ccc0d07e.jpeg    2022-5-8  ...   
1   safe/files/scamDatabase/logo/61e049352a11a.png   2021-6-23  ...   
2   safe/files/scamDatabase/logo/62432b9d39213.png  2019-12-16  ...   
3  safe/files/scamDatabase/logo/6241c6cb5dadc.jpeg   2021-4-22  ...   
4  safe/files/scamDatabase/logo/61bb58e866760.jpeg   2018-1-15  ...   

                           discord bug_bounty_program_link  \
0  https://twitter.com/terra_money                    None   
1                             None                    None   
2                             None                    None   
3                             None                    None   
4                             None                    None   

  bug_bounty_program_company audit_code_conf is_verified_source_code  \
0                       None             0.0                     0.0   
1                       None             NaN                     0.0   
2                       None             NaN                     0.0   
3                       None             NaN                     0.0   
4                       None             NaN                     0.0   

  is_public_team              scam_type network  \
0            1.0   {'type': 'Bank Run'}      {}   
1            1.0  {'type': 'Exit Scam'}      {}   
2            0.0  {'type': 'Exit Scam'}      {}   
3            0.0  {'type': 'Exit Scam'}      {}   
4            0.0  {'type': 'Exit Scam'}      {}   

                                        scamNetworks  \
0  [{'networks_id': 1003, 'scam_database_id': 305...   
1  [{'networks_id': 1666600003, 'scam_database_id...   
2  [{'networks_id': 1666600003, 'scam_database_id...   
3  [{'networks_id': 1666600003, 'scam_database_id...   
4  [{'networks_id': 1666600003, 'scam_database_id...   

                                           auditedBy  
0  [{'audit_link': 'safe/files/audit/pdf/CertiK_A...  
1                                                 []  
2                                                 []  
3                                                 []  
4                                                 []  

[5 rows x 31 columns]

In [3]:
REKT_df.tail(5) # visualize last 5 instances of raw data

id      project_name                                        description  \
3050  78      BoobsFinance                                               None   
3051  77        Chad Token  <p>The contract owner could disable the transf...   
3052  76  Generate Finance  <p><span data-sheets-userformat='{"2":15165,"3...   
3053  72      Pill Finance  <p><span data-sheets-userformat='{"2":15165,"3...   
3054  55     SoftdrinkSwap  <p>The team abandoned the project. The team st...   

     name_categories token_name proof_archive_link technical_issue  \
3050            None      BOOBS               None            None   
3051           Token       CHAD               None            None   
3052            None        GEN               None            None   
3053            None      RED-P               None            None   
3054            None  COLA SODA               None            None   

                                   token_address  \
3050  0xf98f73350d083005079b6c3cda9c99cfdb668be7   
3051  0x2f7383de70c972ef8c18565a9da023b08c110f2c   
3052  0x04ad13a645748cee762f11e43386fe2a275885b4   
3053  0x137faad0d13813ef8d4cbbb336f0e01066b2c9b4   
3054  0x740426e58a10ac3b6724f753c8468288248137f9   

                                            logo_link        date  ...  \
3050   safe/files/scamDatabase/logo/60d50974c47c8.png        None  ...   
3051   safe/files/scamDatabase/logo/60d508fa757bf.png    2021-3-8  ...   
3052  safe/files/scamDatabase/logo/60d50819f1be1.jpeg        None  ...   
3053  safe/files/scamDatabase/logo/60d5003a01e5a.jpeg        None  ...   
3054   safe/files/scamDatabase/logo/60df352021e17.png  2020-11-12  ...   

     discord bug_bounty_program_link bug_bounty_program_company  \
3050    None                    None                       None   
3051    None                    None                       None   
3052    None                    None                       None   
3053    None                    None                       None   
3054    None                    None                       None   

     audit_code_conf is_verified_source_code is_public_team  \
3050             NaN                     NaN            NaN   
3051             NaN                     NaN            NaN   
3052             NaN                     NaN            NaN   
3053             NaN                     NaN            NaN   
3054             NaN                     NaN            NaN   

                  scam_type network  \
3050         {'type': None}      {}   
3051   {'type': 'Honeypot'}      {}   
3052         {'type': None}      {}   
3053         {'type': None}      {}   
3054  {'type': 'Abandoned'}      {}   

                                           scamNetworks auditedBy  
3050  [{'networks_id': 56, 'scam_database_id': 78, '...        []  
3051  [{'networks_id': 56, 'scam_database_id': 77, '...        []  
3052  [{'networks_id': 56, 'scam_database_id': 76, '...        []  
3053  [{'networks_id': 56, 'scam_database_id': 72, '...        []  
3054  [{'networks_id': 56, 'scam_database_id': 55, '...        []  

[5 rows x 31 columns]

In [4]:
#finding out the shape of the data using "shape" variable: Output (rows, columns)
REKT_df.shape

(3055, 31)

In [5]:
#Printing all the columns present in data
REKT_df.columns

Index(['id', 'project_name', 'description', 'name_categories', 'token_name',
       'proof_archive_link', 'technical_issue', 'token_address', 'logo_link',
       'date', 'proof_link', 'website_link', 'webarchive_link', 'twitter_link',
       'telegram_link', 'our_post_link', 'funds_lost', 'funds_returned',
       'active', 'git_hub', 'git_hub_contract_link', 'discord',
       'bug_bounty_program_link', 'bug_bounty_program_company',
       'audit_code_conf', 'is_verified_source_code', 'is_public_team',
       'scam_type', 'network', 'scamNetworks', 'auditedBy'],
      dtype='object')

In [6]:
#Checking for NaN values present in data
REKT_df.isna().sum() # also REKT_df.isnull().sum()

id                               0
project_name                     0
description                    285
name_categories                770
token_name                     194
proof_archive_link            2815
technical_issue               3051
token_address                  209
logo_link                        0
date                          1876
proof_link                     190
website_link                  2220
webarchive_link               2830
twitter_link                  2518
telegram_link                 2628
our_post_link                 3016
funds_lost                    2286
funds_returned                2995
active                           0
git_hub                       2965
git_hub_contract_link         3028
discord                       2917
bug_bounty_program_link       3053
bug_bounty_program_company    3054
audit_code_conf               3032
is_verified_source_code       2513
is_public_team                2513
scam_type                        0
network             

In [7]:
# Removing spaces at beginning and at the end, if any, from column names
REKT_df.columns = REKT_df.columns.str.strip()

## Dropping Variables That Do Not Assist in Analysis

Most of the variables that we shall drop from our 31 columns are links to external sites, such as web archive, discord, and github. Moreover, these variables contain more than 2900 NaN values out of a total observation count of 3055. However, the variable proof_link could be important if I decide to scrape text data from the linked article about a crypto attack. We shall also get rid of the technical_issue field because it only contains 4 non-NaN values and, more importantly, does not have any insightful use. Therefore, I find it sensible to entirely remove these fields, instead of getting rid of their existing NaN values, for EDA and modeling purposes.

In [8]:
REKT_df.drop(columns={'technical_issue', 'proof_archive_link', 'logo_link', 'website_link', 'twitter_link', 'our_post_link', 'telegram_link', 'git_hub', 'git_hub_contract_link', 'discord', 'bug_bounty_program_link', 'bug_bounty_program_company', 'audit_code_conf'}, inplace=True)

REKT_df.shape

(3055, 18)

In [9]:
REKT_df.columns

Index(['id', 'project_name', 'description', 'name_categories', 'token_name',
       'token_address', 'date', 'proof_link', 'webarchive_link', 'funds_lost',
       'funds_returned', 'active', 'is_verified_source_code', 'is_public_team',
       'scam_type', 'network', 'scamNetworks', 'auditedBy'],
      dtype='object')

## Variable Identification and Typecasting

In [10]:
# A closer look at the data types present in the data
REKT_df.dtypes

id                           int64
project_name                object
description                 object
name_categories             object
token_name                  object
token_address               object
date                        object
proof_link                  object
webarchive_link             object
funds_lost                  object
funds_returned              object
active                       int64
is_verified_source_code    float64
is_public_team             float64
scam_type                   object
network                     object
scamNetworks                object
auditedBy                   object
dtype: object

There are a lot of variables visible at once, so let's narrow this down by looking **at one datatype at once**. We will start with **int64** data type.


### Integer Data Type

In [11]:
#Identifying variables with integer datatype
REKT_df.dtypes[REKT_df.dtypes == 'int64']

id        int64
active    int64
dtype: object

In [12]:
REKT_df['active'].value_counts()

1    3055
Name: active, dtype: int64

Summary:

*    **id** is a unique, nominal code indicating the token/coin associated with the crypto attack. Converting it to category type would not be beneficial due to the large number of unique tokens present in the database. **This variable should be converted to object/string type**. 

*    **active** most probably represents whether the crypto project is currently active in the market. I perused the API documentation to try to find this response variable's significance, but could not. Moreover, it has only taken on one as a value for ALL observations. This would mean that all crypto projects present in the database are still active. We could keep this variable for now, but **converting it to category would be better as it most likely would take on two values, either 1 (active) or 0 (inactive).**  

In [13]:
#Converting active to category datatype
REKT_df['active'] = REKT_df['active'].astype('category')
REKT_df['id'] = REKT_df['id'].astype('object')

REKT_df.dtypes

id                           object
project_name                 object
description                  object
name_categories              object
token_name                   object
token_address                object
date                         object
proof_link                   object
webarchive_link              object
funds_lost                   object
funds_returned               object
active                     category
is_verified_source_code     float64
is_public_team              float64
scam_type                    object
network                      object
scamNetworks                 object
auditedBy                    object
dtype: object

### Object Data Type

In [14]:
#Identifying variables with object datatype
REKT_df.dtypes[REKT_df.dtypes == 'object']

id                 object
project_name       object
description        object
name_categories    object
token_name         object
token_address      object
date               object
proof_link         object
webarchive_link    object
funds_lost         object
funds_returned     object
scam_type          object
network            object
scamNetworks       object
auditedBy          object
dtype: object

*    **Variables like 'date', 'funds_lost', and 'funds_returned' are of type object**. This means that **Pandas was not able to recognise the datatype** of these four variables. Therefore, we shall convert these aforementioned object data type variables to their respective datatype

In [15]:
REKT_df['funds_lost'].head(10)

0    40000000000
1     3600000000
2     2900000000
3     2000000000
4     2000000000
5     1000000000
6      625000000
7      602189570
8      534000000
9      473000000
Name: funds_lost, dtype: object

In [16]:
REKT_df['funds_returned'].head(10)

0         None
1         None
2         None
3         None
4     56000000
5         None
6    155800000
7    602189570
8    534000000
9         None
Name: funds_returned, dtype: object

In [17]:
#Converting the object data type variables to their respective datatype
REKT_df['funds_lost'] = REKT_df['funds_lost'].astype('float64')
REKT_df['funds_returned'] = REKT_df['funds_returned'].astype('float64')
REKT_df['date'] = pd.to_datetime(REKT_df['date'])

*    As can be seen above, we have converted 'id' and 'store_and_fwd_flag' to **category**. Due to the 'id' variable's format, we shall have to modify all of its values by removing the redundant 'id' part present in all values. We will also keep the 'id' variable as 'category', **since 'id' represents a unique identifier for each trip and most probably won't be used for analysis**. We change both 'pickup_datetime' and 'dropoff_datetime' to **datetime datatypes**. We will further investigate the datatime datatypes and extract more information from them.

In [18]:
REKT_df.dtypes

id                                 object
project_name                       object
description                        object
name_categories                    object
token_name                         object
token_address                      object
date                       datetime64[ns]
proof_link                         object
webarchive_link                    object
funds_lost                        float64
funds_returned                    float64
active                           category
is_verified_source_code           float64
is_public_team                    float64
scam_type                          object
network                            object
scamNetworks                       object
auditedBy                          object
dtype: object

### Float Data Type

In [19]:
# Identifying variables with float datatype
REKT_df.dtypes[REKT_df.dtypes == 'float64']

funds_lost                 float64
funds_returned             float64
is_verified_source_code    float64
is_public_team             float64
dtype: object

In [20]:
REKT_df['is_verified_source_code'].head(10)

0    0.0
1    0.0
2    0.0
3    0.0
4    0.0
5    0.0
6    0.0
7    0.0
8    0.0
9    0.0
Name: is_verified_source_code, dtype: float64

In [21]:
REKT_df['is_public_team'].head(10)

0    1.0
1    1.0
2    0.0
3    0.0
4    0.0
5    0.0
6    0.0
7    0.0
8    0.0
9    0.0
Name: is_public_team, dtype: float64

*    ***The funds variables should be float types. However, the is_verified_source_code and is_public_team variables take on the values of either 0 or 1. Hence, we convert them to category type.***

In [22]:
REKT_df['is_verified_source_code'] = REKT_df['is_verified_source_code'].astype('category')
REKT_df['is_public_team'] = REKT_df['is_public_team'].astype('category')

REKT_df.dtypes

id                                 object
project_name                       object
description                        object
name_categories                    object
token_name                         object
token_address                      object
date                       datetime64[ns]
proof_link                         object
webarchive_link                    object
funds_lost                        float64
funds_returned                    float64
active                           category
is_verified_source_code          category
is_public_team                   category
scam_type                          object
network                            object
scamNetworks                       object
auditedBy                          object
dtype: object

In [23]:
REKT_df.head()

id   project_name                                        description  \
0  3058  Terra Classic  <p><strong>Quick Summary</strong></p><p>A comp...   
1  2762      Africrypt  <p><strong>Quick Summary</strong></p><p>Ameer ...   
2  2878      PlusToken  <p><strong>Quick Summary</strong></p><p>The pe...   
3  2861         Thodex  <p><strong>Quick Summary</strong></p><p>Thodex...   
4  2735     BitConnect  <p><strong>Quick Summary</strong></p><p>BitCon...   

              name_categories  token_name  \
0                  Stablecoin  LUNC, USTC   
1                        CeFi        None   
2                        CeFi        None   
3                        CeFi        None   
4  Borrowing and Lending,CeFi         BCC   

                                       token_address       date  \
0  0x7e43d25EaD96B1058f671F6690ea705BA2C7e5B9, 0x... 2022-05-08   
1                                               None 2021-06-23   
2                                               None 2019-12-16   
3                                               None 2021-04-22   
4                                               None 2018-01-15   

                                          proof_link  \
0  https://www.nansen.ai/research/on-chain-forens...   
1  https://finance.yahoo.com/news/africrypt-bitco...   
2  https://cointelegraph.com/news/vanuatu-extradi...   
3  https://www.cnbc.com/2021/04/23/bitcoin-btc-ce...   
4                                               None   

                                     webarchive_link    funds_lost  \
0                                               None  4.000000e+10   
1  https://web.archive.org/web/20200921145240/htt...  3.600000e+09   
2  https://web.archive.org/web/20220125033211/htt...  2.900000e+09   
3  https://web.archive.org/web/20220405133149/htt...  2.000000e+09   
4  https://web.archive.org/web/20220426131429/htt...  2.000000e+09   

   funds_returned active is_verified_source_code is_public_team  \
0             NaN      1                     0.0            1.0   
1             NaN      1                     0.0            1.0   
2             NaN      1                     0.0            0.0   
3             NaN      1                     0.0            0.0   
4      56000000.0      1                     0.0            0.0   

               scam_type network  \
0   {'type': 'Bank Run'}      {}   
1  {'type': 'Exit Scam'}      {}   
2  {'type': 'Exit Scam'}      {}   
3  {'type': 'Exit Scam'}      {}   
4  {'type': 'Exit Scam'}      {}   

                                        scamNetworks  \
0  [{'networks_id': 1003, 'scam_database_id': 305...   
1  [{'networks_id': 1666600003, 'scam_database_id...   
2  [{'networks_id': 1666600003, 'scam_database_id...   
3  [{'networks_id': 1666600003, 'scam_database_id...   
4  [{'networks_id': 1666600003, 'scam_database_id...   

                                           auditedBy  
0  [{'audit_link': 'safe/files/audit/pdf/CertiK_A...  
1                                                 []  
2                                                 []  
3                                                 []  
4                                                 []

### datetime Data Type

*    ***We shall now extract important time-based features for better EDA experience***

In [24]:
# create time based features for pickup_datetime 
REKT_df['month_of_attack'] = REKT_df.date.dt.month
REKT_df['day_of_week_of_attack'] = REKT_df.date.dt.dayofweek
REKT_df['day_of_year_of_attack'] = REKT_df.date.dt.dayofyear

In [25]:
REKT_df.head()

id   project_name                                        description  \
0  3058  Terra Classic  <p><strong>Quick Summary</strong></p><p>A comp...   
1  2762      Africrypt  <p><strong>Quick Summary</strong></p><p>Ameer ...   
2  2878      PlusToken  <p><strong>Quick Summary</strong></p><p>The pe...   
3  2861         Thodex  <p><strong>Quick Summary</strong></p><p>Thodex...   
4  2735     BitConnect  <p><strong>Quick Summary</strong></p><p>BitCon...   

              name_categories  token_name  \
0                  Stablecoin  LUNC, USTC   
1                        CeFi        None   
2                        CeFi        None   
3                        CeFi        None   
4  Borrowing and Lending,CeFi         BCC   

                                       token_address       date  \
0  0x7e43d25EaD96B1058f671F6690ea705BA2C7e5B9, 0x... 2022-05-08   
1                                               None 2021-06-23   
2                                               None 2019-12-16   
3                                               None 2021-04-22   
4                                               None 2018-01-15   

                                          proof_link  \
0  https://www.nansen.ai/research/on-chain-forens...   
1  https://finance.yahoo.com/news/africrypt-bitco...   
2  https://cointelegraph.com/news/vanuatu-extradi...   
3  https://www.cnbc.com/2021/04/23/bitcoin-btc-ce...   
4                                               None   

                                     webarchive_link    funds_lost  ...  \
0                                               None  4.000000e+10  ...   
1  https://web.archive.org/web/20200921145240/htt...  3.600000e+09  ...   
2  https://web.archive.org/web/20220125033211/htt...  2.900000e+09  ...   
3  https://web.archive.org/web/20220405133149/htt...  2.000000e+09  ...   
4  https://web.archive.org/web/20220426131429/htt...  2.000000e+09  ...   

   active is_verified_source_code is_public_team              scam_type  \
0       1                     0.0            1.0   {'type': 'Bank Run'}   
1       1                     0.0            1.0  {'type': 'Exit Scam'}   
2       1                     0.0            0.0  {'type': 'Exit Scam'}   
3       1                     0.0            0.0  {'type': 'Exit Scam'}   
4       1                     0.0            0.0  {'type': 'Exit Scam'}   

  network                                       scamNetworks  \
0      {}  [{'networks_id': 1003, 'scam_database_id': 305...   
1      {}  [{'networks_id': 1666600003, 'scam_database_id...   
2      {}  [{'networks_id': 1666600003, 'scam_database_id...   
3      {}  [{'networks_id': 1666600003, 'scam_database_id...   
4      {}  [{'networks_id': 1666600003, 'scam_database_id...   

                                           auditedBy month_of_attack  \
0  [{'audit_link': 'safe/files/audit/pdf/CertiK_A...             5.0   
1                                                 []             6.0   
2                                                 []            12.0   
3                                                 []             4.0   
4                                                 []             1.0   

   day_of_week_of_attack  day_of_year_of_attack  
0                    6.0                  128.0  
1                    2.0                  174.0  
2                    0.0                  350.0  
3                    3.0                  112.0  
4                    0.0                   15.0  

[5 rows x 21 columns]

In [26]:
REKT_df.tail()

id      project_name                                        description  \
3050  78      BoobsFinance                                               None   
3051  77        Chad Token  <p>The contract owner could disable the transf...   
3052  76  Generate Finance  <p><span data-sheets-userformat='{"2":15165,"3...   
3053  72      Pill Finance  <p><span data-sheets-userformat='{"2":15165,"3...   
3054  55     SoftdrinkSwap  <p>The team abandoned the project. The team st...   

     name_categories token_name                               token_address  \
3050            None      BOOBS  0xf98f73350d083005079b6c3cda9c99cfdb668be7   
3051           Token       CHAD  0x2f7383de70c972ef8c18565a9da023b08c110f2c   
3052            None        GEN  0x04ad13a645748cee762f11e43386fe2a275885b4   
3053            None      RED-P  0x137faad0d13813ef8d4cbbb336f0e01066b2c9b4   
3054            None  COLA SODA  0x740426e58a10ac3b6724f753c8468288248137f9   

           date                                         proof_link  \
3050        NaT                                               None   
3051 2021-03-08                                               None   
3052        NaT  https://twitter.com/WARONRUGS/status/136972604...   
3053        NaT  https://twitter.com/RugSteemer/status/13717216...   
3054 2020-11-12                                               None   

               webarchive_link  funds_lost  ...  active  \
3050                      None         NaN  ...       1   
3051                      None         NaN  ...       1   
3052                      None         NaN  ...       1   
3053  https://archive.ph/AfyEo         NaN  ...       1   
3054                      None         NaN  ...       1   

     is_verified_source_code is_public_team              scam_type network  \
3050                     NaN            NaN         {'type': None}      {}   
3051                     NaN            NaN   {'type': 'Honeypot'}      {}   
3052                     NaN            NaN         {'type': None}      {}   
3053                     NaN            NaN         {'type': None}      {}   
3054                     NaN            NaN  {'type': 'Abandoned'}      {}   

                                           scamNetworks auditedBy  \
3050  [{'networks_id': 56, 'scam_database_id': 78, '...        []   
3051  [{'networks_id': 56, 'scam_database_id': 77, '...        []   
3052  [{'networks_id': 56, 'scam_database_id': 76, '...        []   
3053  [{'networks_id': 56, 'scam_database_id': 72, '...        []   
3054  [{'networks_id': 56, 'scam_database_id': 55, '...        []   

     month_of_attack  day_of_week_of_attack  day_of_year_of_attack  
3050             NaN                    NaN                    NaN  
3051             3.0                    0.0                   67.0  
3052             NaN                    NaN                    NaN  
3053             NaN                    NaN                    NaN  
3054            11.0                    3.0                  317.0  

[5 rows x 21 columns]

In [27]:
REKT_df.dtypes

id                                 object
project_name                       object
description                        object
name_categories                    object
token_name                         object
token_address                      object
date                       datetime64[ns]
proof_link                         object
webarchive_link                    object
funds_lost                        float64
funds_returned                    float64
active                           category
is_verified_source_code          category
is_public_team                   category
scam_type                          object
network                            object
scamNetworks                       object
auditedBy                          object
month_of_attack                   float64
day_of_week_of_attack             float64
day_of_year_of_attack             float64
dtype: object

In [28]:
REKT_df.describe()

funds_lost  funds_returned  month_of_attack  day_of_week_of_attack  \
count  7.690000e+02    6.000000e+01      1179.000000            1179.000000   
mean   7.915904e+07    3.961377e+07         7.302799               2.572519   
std    1.455806e+09    1.155030e+08         3.109340               1.948956   
min    1.600000e+01    4.440000e+03         1.000000               0.000000   
25%    4.776500e+04    4.911965e+05         5.000000               1.000000   
50%    3.510640e+05    2.758203e+06         8.000000               3.000000   
75%    3.000000e+06    7.517699e+06         9.000000               4.000000   
max    4.000000e+10    6.021896e+08        12.000000               6.000000   

       day_of_year_of_attack  
count            1179.000000  
mean              207.167091  
std                95.505989  
min                 1.000000  
25%               146.000000  
50%               239.000000  
75%               263.000000  
max               366.000000

*   Now we are done breaking down the date of attack datetime variable into granular forms, as seen above, which will **help us better analyze crypto crimes.** We are also done with the Variable identification and typecasting process and will now start the Univariate Analysis portion of the EDA, followed by Bivariate Analysis, and, lastly, Multivariate Analysis.

In [29]:
REKT_df.to_csv("../../data/Clean Data/REKT_Database_Clean_Python.csv")
